### PR TITLE
fix: add timeout when fetching data

### PR DESCRIPTION
**Pull request by yao-peijun**
[Notion link](https://www.notion.so/sally-wallet/Handling-FE-Errors-when-Offline-b296a4620ca04f469c3d144346ba9670) 

This PR add timeout when fetching data to fix infinite loading screen when offline.

- [X] I've kept this PR as small as possible (~600 lines) by splitting it into PRs with manageable chunks of code
- [X] I've requested reviews from 1-2 reviewers
- [ ] I've added [accessibility IDs](https://www.notion.so/dc5f4ce910f7431c84344ac79344e9f5?v=d487366816834dd4ab8dc12e0b5928f3) to my components
- [ ] I've added/updated unit/integration tests
- [ ] I've added/updated [e2e tests](https://www.notion.so/e2e-Documentation-a096d3e0bf75485e85ad692af8371ef3) for at least the happy flow
- [ ] I've added [Chinese translations for i18n setup](https://www.notion.so/Translations-Dev-Guide-8c1da838982a4f59a386ec96ac6780c8)
- [ ] I've added documentation in README/[Notion](https://www.notion.so/82b92fb1007640328dab9582c0a3694e?v=3b6ce48202cf40ad8450553799b13146)

### CODE REVIEW

File: src/services/helpers.ts [hunk 1]
```diff
@@ -23,14 +23,27 @@ export class NetworkError extends Error {
   }
 }
 
+const timeoutAfter = (seconds: number): Promise<Error> => {
```
Comment (dalsontws) [inline src/services/helpers.ts:26]: just curious, was looking [around](https://stackoverflow.com/questions/42147733/doing-a-timeout-error-with-fetch-react-native/54208009) and `AbortController` was talked about as a recommended way to abort `fetch`,  maybe you can share why you decided to use `Promise.race` and a custom timeout instead?

Comment (yao-peijun) [reply src/services/helpers.ts:26]: abortController() aborts the connections and Promise.race returns fetch data or throw specify error after x seconds which ever that is earlier. 

updated the code to abort connection properly after timeout as recommended.

File: src/services/helpers.ts [hunk 2]
```diff
@@ -23,14 +23,40 @@ export class NetworkError extends Error {
...
 ): Promise<T> {
   let response: Response;
+  /*
+   * AbortController() is supported globally for Jest with Node v15 and Jest v27 and above.
+   * Thus, logic is added to check if AbortController is valid for Jest testing purpose only.
+   * TODO: Remove the logic check when node and jest version is compatible.
+   */
   try {
-    response = await fetch(requestInfo, init);
+    const controller = typeof AbortController
```
Comment (dalsontws) [inline src/services/helpers.ts:46]: why do we need the check here for `typeof AbortController`? can we just `const controller = new AbortController();`?

Comment (yao-peijun) [reply src/services/helpers.ts:46]: Logic added mainly for unit-testing purpose only as AbortController() is not support globally for Node v12.22.1 and Jest v26.6.3 which cause the unit-testing to fail in [92aea07](https://github.com/rationally-app/mobile-application/pull/569/commits/92aea0710ffb30e175604da1ae550efab08751ab). AbortController() is supported globally from Node v15 and Jest v27. [reference](https://github.com/facebook/jest/pull/11182)

**Review by dalsontws — APPROVED** [review]
thanks for explaining and taking the time to research on this, LGTM!

**Pull request merged**